google/cloud/internal/curl_impl.cc

```diff
@@ -241,21 +243,22 @@ CurlImpl::~CurlImpl() {
 std::size_t CurlImpl::WriteAllBytesToSpillBuffer(void* ptr, std::size_t size,
                                                  std::size_t nmemb) {
   // The size of the spill buffer is the same as the max size curl
-  // will write. libcurl should never give us more bytes than we have set in
-  // CURLOPT_BUFFERSIZE. If it does, this is an error in libcurl, and we
-  // cannot reliably recover from it.
-  if (size * nmemb > spill_.max_size()) {
-    GCP_LOG(FATAL) << absl::StrCat("libcurl attempted to write ",
-                                   std::to_string(size * nmemb),
-                                   "bytes into spill buffer size ",
-                                   std::to_string(spill_.max_size()))
+  // will write in one invocation. libcurl should never give us more bytes than
+  // we have set in CURLOPT_BUFFERSIZE, on a single write. However, libcurl can
+  // give us fewer bytes which can result in multiple calls to this function on
+  // the initial read.
+  if (size * nmemb > spill_.max_size() - spill_offset_) {
+    GCP_LOG(FATAL) << absl::StrCat(
+                          "libcurl attempted to write ",
+                          std::to_string(size * nmemb),
+                          " bytes into spill buffer with remaining capacity ",
+                          std::to_string(spill_.max_size() - spill_offset_))
                    << "\n";
   }
-  // As this function is only called on the first Read, the spill buffer should
-  // be empty so we can always copy the full contents of `ptr` into the spill
-  // buffer.
-  std::memcpy(spill_.data(), ptr, size * nmemb);
-  spill_offset_ = size * nmemb;
+  // As this function can be called multiple times on the first Read, the spill
+  // buffer may be written to several times before it is full.
+  std::memcpy(spill_.data() + spill_offset_, ptr, size * nmemb);
+  spill_offset_ += size * nmemb;
```

Comment:
Just an observation: this is starting to look similar to how we write into the application's buffer.